### PR TITLE
Hx711

### PR DESCRIPTION
**Pull request by wiiggee1**
# Pull reequest for the HX711 driver. 

This is currently working but needs calibration. Please look over and review the code, before merge to main.

### CODE REVIEW

File: src/drivers/hx711.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    delay::DelayNs,
+    digital::{InputPin, OutputPin},
+};
+use rtic_monotonics::{
+    nrf::timer::{ExtU64, Timer0 as Mono},
+    systick::fugit::Duration,
+    Monotonic,
+};
+
+pub type Instant = <Mono as rtic_monotonics::Monotonic>::Instant;
```
Comment (ivajon) [inline src/drivers/hx711.rs]: Missing docs for the instant

File: src/drivers/hx711.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+/// The gain act as the 'stop index' for the pulses.
+/// Where:
+/// - Gain(128) => 25 bit pulses.
+/// - Gain(32) => 26 bit pulses.
+/// - Gain(64) => 27 bit pulses.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy)]
+pub enum Gain {
+    Apply128 = 25,
```
Comment (ivajon) [inline src/drivers/hx711.rs:62]: Move doc comments to here instead. Each public enum variant should have doc comments

File: src/drivers/hx711.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+};
+
+pub type Instant = <Mono as rtic_monotonics::Monotonic>::Instant;
+
+
+/// Hx711 driver struct.
+pub struct Hx711Driver<PINOUT: OutputPin + Send, PININ: InputPin + Send> {
+    pub pd_sck: PINOUT,
+    pub dout: PININ,
+    pub dout_data: i32,
```
Comment (ivajon) [inline src/drivers/hx711.rs:35]: Docs

File: src/drivers/hx711.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+/// T1: DOUT falling edge to `PD_SCK` rising edge.
+/// T2: `PD_SCK` rising edge to DOUT data ready.
+/// T3: `PD_SCK` high time.
+/// T4: `PD_SCK` low time.
+pub enum DataTiming {
+    /// MIN 0.1 µs .
+    T1,
+    /// MAX 0.1 µs.
```
Comment (ivajon) [inline src/drivers/hx711.rs]: Either this or the From<DataTiming> is incorrect

File: src/drivers/hx711.rs [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        match val {
+            Gain::Apply128 => 25,
+            Gain::Apply32 => 26,
+            Gain::Apply64 => 27,
+        }
+    }
+}
+
+impl<PINOUT: OutputPin + Send, PININ: InputPin + Send> Hx711Driver<PINOUT, PININ> {
+    pub const CALIBRATION_CONST: i32 = 3065448 as i32;
```
Comment (ivajon) [inline src/drivers/hx711.rs]: Docs

File: src/drivers/hx711.rs [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    /// Initialize and creates a new HX711 driver instance.
+    pub const fn init(sck: PINOUT, digital_out: PININ, gain_amount: Gain) -> Self {
+        Self {
+            pd_sck: sck,
+            dout: digital_out,
```
Comment (ivajon) [inline src/drivers/hx711.rs:176]: Add these to [ignore.txt](.github/ignore.txt)

Comment (ivajon) [reply src/drivers/hx711.rs:176]: .github/ignore.txt

File: src/drivers/hx711.rs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        Mono::delay(DataTiming::T4.into()).await;
+
+        read_done
+    }
+
+    /// This method, would apply the gain for the next data
+    /// sample period. 
+    fn next_conversion_gain(&mut self, gain_amount: Gain) -> u8 {
+        self.gain = gain_amount;
+        let stop_index: u8 = match gain_amount {
```
Comment (ivajon) [inline src/drivers/hx711.rs]: Replace with .into() no need for the match

File: src/drivers/hx711.rs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            Gain::Apply64 => Gain::Apply64.into(),
+        };
+        stop_index - 1
+    }
+
+    /// Reads a 24-bit full period data. Where data is 
+    /// shifted out on the `dout` pin
+    pub async fn read_full_period(&mut self, next_gain: Gain) -> i32 {
+        let stop_count = self.next_conversion_gain(next_gain);
+        Mono::delay(DataTiming::T1.into()).await;
```
Comment (ivajon) [inline src/drivers/hx711.rs]: We miiiight want this to be a generic argument like we did in the angle sensor, the syntax is uglier but works for more boards.

File: src/drivers/hx711.rs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        // Can I use i32 directly (two's complement) or do I need to wrap,
+        // and convert the u32 to i32?
+        //let data_sample = unsafe { i32::try_from(data_in).unwrap_unchecked() };
+        
+        let data_sample = if data_in >= 0x800000 && data_in < 0x7FFFFF {
+            //data_in.wrapping_neg() - Self::CALIBRATION_CONST
+            data_in
+
+        } else {
+            data_in
```
Comment (ivajon) [inline src/drivers/hx711.rs:264]: Is this correct? Both paths return the same value

File: src/drivers/hx711.rs [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    Apply128 = 25,
+    Apply32 = 26,
+    Apply64 = 27,
+}
+
+/// T1: DOUT falling edge to `PD_SCK` rising edge.
+/// T2: `PD_SCK` rising edge to DOUT data ready.
+/// T3: `PD_SCK` high time.
+/// T4: `PD_SCK` low time.
+pub enum DataTiming {
```
Comment (ivajon) [inline src/drivers/hx711.rs]: I do not think that this needs to be public.

File: src/drivers/hx711.rs [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    nrf::timer::{ExtU64, Timer0 as Mono},
+    systick::fugit::Duration,
+    Monotonic,
+};
+
+pub type Instant = <Mono as rtic_monotonics::Monotonic>::Instant;
+
+
+/// Hx711 driver struct.
+pub struct Hx711Driver<PINOUT: OutputPin + Send, PININ: InputPin + Send> {
```
Comment (ivajon) [inline src/drivers/hx711.rs:28]: Make the fields that do not need to be public in to private fields

**Review by ivajon — COMMENTED** [review]
Looks pretty good in general, we should make the fields that do not need to be public in to private fields.

File: src/drivers/hx711.rs (hunk 11 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs:28]: Fixed!

File: src/drivers/hx711.rs (hunk 9 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs:264]: Typo fixed it!

File: src/drivers/hx711.rs (hunk 8 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs]: Added so its generic!

File: src/drivers/hx711.rs (hunk 7 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs]: Fixed!

File: src/drivers/hx711.rs (hunk 5 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs]: Added docs.

File: src/drivers/hx711.rs (hunk 4 above)
Comment (wiiggee1) [reply src/drivers/hx711.rs]: Please elaborate more on this.

Comment (wiiggee1) [reply src/drivers/hx711.rs]: Hopefully I made it more clear now!

**Pull request merged**